### ISSUE TITLE
Events never flush after a crash (2.8.2)

### ISSUE THREAD

**Issue by idStar** [mixpanel/mixpanel-iphone#303]
I've noticed consistently while building/testing my crash reporting, that event activity is no longer sent to Mixpanel once a crash has occurred.

##### How to Reproduce
1. Ensure normal activity in your app is sending a steady stream of event activity to Mixpanel.
2. Raise an exception in your code, purposely, that you don't catch. In my case, I was using [FCYAsserts](https://github.com/fcy/FCYAsserts), but you could use [Marcus Zarra's ZAssert()](http://www.cimgf.com/2010/05/02/my-current-prefix-pch-file/) or other similar, assertion macros that raise exceptions.
3. Without rebuilding your app, re-launch it and choose a different pathway through it that won't crash, but which should generate additional tracking events. 

##### Discussion

It's when doing Step 3, even hours after the crash and subsequent app activity, that no matter what I did, no events would come in from the app.

I would have to lay in a fresh copy of the app by building from Xcode onto the device, to get the Mixpanel queue to be 'reset'. However, previously _jammed_ events were lost. I would never see them show up on the Mixpanel dashboard.

I suspect that after a crash, the Mixpanel archive of events is in some corrupted, unrecoverable state.

I have enough logic in my app that I know if a launch was preceded by a crash. At such junctures, is there some way I can force a reset of the Mixpanel cache, so I can get back to a reliable state?

**Comment by samgreen** [comment]
This is a duplicate of #251

**Comment by samgreen** [comment]
This should be resolved with #306

**Comment by patedit** [comment]
Solution proposed on #306 works with current Crashlytics library and both exception handlers can co-exist at the same time. It will be included in the next release v2.9.2. Meanwhile, feel free to checkout #306 and use it.
We're not able to reproduce the lost of events after a crash following your steps. Could you give some more information?

**Comment by idStar** [comment]
@patedit I'm now putting together a sample app to recreate this phenomenon. Sorry for the delay. As mentioned earlier, I'm starting to think this is going to turn out to be the combination of {HockeyApp, Mixpanel} whereas {Crashlytics, Mixpanel} seem to now place nice vis-a-vis top level uncaught exception handlers. Will post findings within a day or so (have to setup profiles for each service for my demo project to illustrate this).

**Comment by idStar** [comment]
##### Update
* Creating a fresh test app (attached) with Cocoapods, and using the latest official HockeySDK-source pod and Mixpanel cocoapod worked well. 
* The event that was created right before the crash (of the button tap that triggered a manual crash) was itself lost (I can live with that)
* Subsequent events after a re-launch however, continued to resume logging (i.e. No jamming)

I was originally going to build using the 306 branch with the exception handler patch, but since my own project is now on carthage, this older branch didn’t play well with Carthage. Of course, subsequent releases of the Mixpanel SDK do work well with Carthage (I might just temporarily integrate manually).

Is it possible you can update the Pull Request 306 branch, rebasing from master?

When I pointed my real app to the last Mixpanel SDK in master and the latest HockeySDK framework (pulled in via Carthage), the jamming was still there.

###### However:

1. I’m not yet testing with the exception handler fix in branch 306, which may have fixed this, and which was the original item you wanted me to test.
2. That the test app works fine, means that some of the optional delegate methods I’ve implemented for the HockeyApp crash reporting response may be what’s causing the problem, as my test app didn’t do anything but use the stock HockeyApp stack.
3. Since my earlier tests with branch 306 in the summer, when I had been using Cocoapods also didn’t bear fruit, I strongly suspect that once someone starts customizing (implementing) crash released delegate methods with HockeyApp, something goes awry.

I’m going to experiment with stripping out robust Hockey App support from my real app, to see if that allows the current master branch of the Mixpanel SDK to work.

**Comment by samgreen** [comment]
Hey @idStar Thanks so much for the very informative comments. I think the earliest we can cut a release would be next month (perhaps mid Jan). I can however rebase these changes from master and put it on a feature branch if that's helpful, although it might be easier for you to just fork the repo.

If you want a feature branch let me know and I can do that ASAP. Otherwise we'll release in Jan.

**Comment by idStar** [comment]
Thanks @samgreen. A feature branch with 306 rebased would be great, but if you can't get to it, I'll fork it and try to get that working myself with Carthage. 

Incidentally, commenting out some of the extra support HockeyApp provides _did_ allow events to go through after a crash. I'm now isolating which bits of API are problematic. It's tedious, but I will report back here when I have it narrowly isolated. 

At that point, there may be a conversation warranted with the HockeyApp folks re: interoperability when `BITCrashManagerDelegate` methods are implemented.

**Comment by samgreen** [comment]
Amazing work on this. Send me an email sam.green@mixpanel.com!

**Comment by idStar** [comment]
Update:

Looks like it is the HockeyApp SDK `BITCrashManagerDelegate` method:

```
- (NSString *)applicationLogForCrashManager:(BITCrashManager *)crashManager
```

which creates the conflict. Implementing this optional delegate method for the HockeySDK causes this silent problem.

Even a trivial implementation, that just returns a short string, causes subsequent Mixpanel events to no longer get sent out, ever.

This is using the latest HockeyApp release (v3.8.5) and pointing to the master branch (as of today) for the Mixpanel iOS SDK.

As a final confirmation, I'll implement that delegate method in my trivially simple test app, to see if that problem is repeated. That'll rule out other possible confounding complexity from my non-trivial, real-world app.

**Comment by idStar** [comment]
In my trivial, simple test app, implementing this HockeyApp SDK method did _not_ cause a problem. So, I'm going to enhance my test app to implement a more robust HockeyApp delegate. I suspect we're dealing with a threshold of too much activity around a crash, that other frameworks (like Mixpanel) are unable to get in a word edgewise.

However, these methods generally fire or re-launch _after_ a crash, so it's still a mystery as to why that would even matter. 

In any event, I hope to demonstrate this in a more robust sample app, which I am working on now.

**Comment by idStar** [comment]
Update:
* I added in a lot of paraphernalia to the sample app to try to start approximating the interplay of various services on startup of the app
* I still could not create the problem evidenced by my real app
* This is all using the current `master` branch of the Mixpanel SDK; not even using the uncaught exception handler fix of Issue #306.
* Since commenting out one Hockey App delegate method allowed Mixpanel events to go through on relaunch after a crash in my real app, I started re-ordering some initialization code.

In the end, I got everything to work in my _real_ app, in combination with the `master` branch of code of the Mixpanel SDK, by eagerly (instead of lazily) initializing my adapter class for the `Mixpanel` shared instance.

There's some delicate timing here, but I can't definitely explain why the fix I stumbled upon, worked.

I do thing the refactoring in Issue #306 is worth merging in, as well as looking at the main crash management services/SDKs (Crashlytics, HockeyApp, etc.) to at least audit where conflicts or race conditions may arise.
